### ISSUE TITLE
Modal: Groot onderhoud (dialog element, responsive bugfix, geen ModalControllers meer, fullscreen support)

### ISSUE THREAD

**Issue by tfrijsewijk** [dso-toolkit/dso-toolkit#2223]
# Issue achtergrond

In #2092 is een responsive bug geintroduceerd. Daarnaast is een deel van de scope niet uitgewerkt. Het gaat over het positioneren van de Modal. Deze positionering is achterhaald doordat we het `<dialog>` element gaan inzetten. Een `<dialog>` laat zich slecht positioneren. Er is afgesproken om nadat we het `<dialog>` element hebben ingezet uitzoeken of we de Modal nog kunnen laten positioneren.

De bug staat en moet in dit issue worden opgelost. De afmetingen uit #2092 dienen als basis voor de maximale afmetingen van het modal.

In dit issue worden ook #2152 en #2215 opgepakt.

## `<dialog>`

Overgaan op gebruik van het `<dialog>`-element voor _alle_ modals. Dat betekent:

- z-index overbodig, `dialog` wordt met `showModal()` in de `#top-layer` geplaatst, boven alles.
- support is goed: https://caniuse.com/?search=dialog
- krijgt vanuit de browser een stijlbare `::backdrop`, dus geen aparte overlay meer nodig
- focus trap (test: focus is trapped, behalve naar de browser url bar; dat laatste is 'by design'. Test het hier: https://codepen.io/picturit/live/gOQRJyq )
- a11y support: NVDA test (Windows 11, chrome + NVDA - Parallels) op https://codepen.io/picturit/live/gOQRJyq door mij (Hans) uitgevoerd met positief resultaat: modal wordt direct bij openen helemaal voorgelezen, en bij sluiten keert de focus terug op de knop die het modal geopend heeft.
- proeftuintje code: https://codepen.io/picturit/pen/gOQRJyq
- zo te zien kan je nog steeds de pagina eronderdoor scrollen; dat zou je dan indien gewenst apart uit moeten zetten met een `overflow: hidden` zolang het modal getoond wordt.

### Werkzaamheden

- Voor de HTML/CSS implementatie de buitenste `<div>` vervangen door `<dialog>`.
- De documentatie voor het HTML/CSS component uitbreiden met instructies hoe een dialog kan worden geopend.
- In het Web Component wordt het root element een  `<dialog>`. Tijdens het instantieren van `<dso-modal>` wordt de modal direct geopend met `HTMLDialogElement.showModal()`.
- Eigen backdrop eraf, styling naar `::backdrop`
- Zorgen voor een accessible name dmv. `aria-labelledby`, zoals nu al gebeurt.

Zonder `<dialog>` kent een Modal een aantal uitdagingen (toegankelijkheid, als eerst in DOM, tab index, etc). Hiervoor zijn zng. ModalControlls gemaakt. Core, React en Angular hebben ieder hun eigen implementatie. Deze zijn door `<dialog>` overbodig en kunnen worden verwijderd.

## Bug: Mobile viewport Modal is te kort.

Een Modal op mobile viewport pakt niet de maximale ruimte:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/064a5e02-bfef-40f7-9839-c818d027f259)

## API uitbreiding: Fullscreen

```html
<dso-modal fullscreen>
```

Deze boolean attribute zorgt ervoor dat de Modal volledig dekkend in viewport is. Dezelfde positionering voor de elementen als titel en knoppen blijven gehanteerd.

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/19a31207-78ba-436a-a1ad-bf85050a78e1)

## Checklist

- [ ] `<dialog>` in HTML/CSS implementatie
- [ ] Documentatie voor Web Component en HTML/CSS component bijgewerkt.
- [ ] `<dialog>` in Web Component met automatisch `HTMLDialogElement.showModal()`
- [ ] `fullscreen` (Modal model, templates: HTML/CSS, Core & React, styling en Storybook controls)
- [ ] ModalControllers (Core, React & Angular) verwijderd.
- [ ] Responsive bug
- [ ] Schaduwtjes bij Modal met teveel content werken nog (`<dso-scrollable>`).

**Comment by LukeVerlaan** [comment]
@tfrijsewijk Het webcomponent en de html css variant zijn allebei goed met scrollable. Voor html css met `.dso-body` kan je dit niet realiseren. 
Persoonlijk ben ik zoiezo al voorstander van het deprecaten van de html/css modal, omdat dit gewoon niet onderhoudbaar is ivm het webcomponent. Daar is het component te complex voor.

**Comment by LukeVerlaan** [comment]
We gaan proberen om het "uitrekenen"  van de body te vervangen met flex (flex-grow enz) oid.

**Comment by HalloJo** [comment]
Daar ben ik nu flink mee bezig. Het is flink rommelen.

**Comment by HalloJo** [comment]
Heb wat ik tot nu toe heb gepushed op de branch. We komen in de buurt, maar zijn er nog niet; Shaduwtjes vallen nu weg uit de `scrollable` omdat ik `flex` heb toegepast. Daar probeer ik nu iets voor te bedenken.  🔨

**Comment by LukeVerlaan** [comment]
Ik heb de scrollable gefixt.
Wat er nog moet gebeuren:
 - "minimale hoogte modal: bij viewport hoogte < 232 px wordt de modal niet meer kleiner"
 - de vershillende viewport widths

**Comment by tfrijsewijk** [comment]
De scope van dit issue is aangepast. Het ene issue haalde het andere issue in. Bij UX zijn verwachtingen aangepast (geen top-positionering van 40px met max-height tot 40px onder meer).

Daarnaast is het belang van `<dialog>` aangegeven en is UX akkoord 🎉 

@hansgrimm kan jij de scope reviewen/refinen?

**Comment by hansgrimm** [comment]
De fullscreen styling staat erin v.w.b. het webcomponent. de html/css versie kan ik nog niet goed checken, omdat die (nog) niet met `showModal()` afgetrapt is; dat is nodig omdat de presentatie deels leunt op UA-styling. 
Ik heb zelf ook een poging gedaan om die aftrap erin te krijgen, maar zonder succes.
Als iemand hier een list kan verzinnen.. ;)

**Comment by hansgrimm** [comment]
Net gefixt: 
Bug: Mobile viewport Modal is te kort.
Volgens mij hebben we nu alle 3 de issues in de steigers staan; alleen nog de `showModal()` op de html/css versie, en dan kunnen we het af gaan maken.

**Comment by hansgrimm** [comment]
Wat mij betreft staat de styling van core en html/css nu strak, alleen nog even (stress-)testen en ik doe er vooralsnog even een strik omheen.
@LukeVerlaan heb net jouw commits binnengehengeld, maar krijg nu in mijn lokale builld deze errors:
![Scherm­afbeelding 2023-08-10 om 13 27 03](https://github.com/dso-toolkit/dso-toolkit/assets/506527/a82bb38a-d082-4681-b4d6-bd850dc8f213)
De omgeving werkt wel, maar die errors zijn licht verontrustend ;)

**Comment by LukeVerlaan** [comment]
Die errors zie ik ook inderdaad, maar ik heb echt totaal geen idee waarom die opeens tevoorschijn zijn gekomen.

**Comment by hansgrimm** [comment]
Ze gaan ook over iets compleet anders...

**Comment by hansgrimm** [comment]
O ja: ik heb een manier gevonden om een `<dialog>` toch anders te positioneren dan horizontaal/verticaal gecentreerd. de UA-styling maakt gebruik van `margin: auto`, dus je kan de `margin-block-start` gewoon ingeven als je de modal hogerop wil hebben. Nog wel even goed testen, maar dit lijkt te werken.

**Comment by hansgrimm** [comment]
@LukeVerlaan kunnen we hier al een PR van bakken, of moet jij nog iets doen? De styling staat m.i. nu wel strak, maar dat is niet het enige 😉

**Comment by tfrijsewijk** [comment]
De branch build niet vanwege een linterror. En nog even rebasen op master.

**Comment by hansgrimm** [comment]
> De scope van dit issue is aangepast. Het ene issue haalde het andere issue in. Bij UX zijn verwachtingen aangepast (geen top-positionering van 40px met max-height tot 40px onder meer).
> 
> Daarnaast is het belang van `<dialog>` aangegeven en is UX akkoord 🎉
> 
> @hansgrimm kan jij de scope reviewen/refinen?

Je moet me even helpen, want ik zie niet direct wat er qua scope is aangepast.
De linterror  zit zo te zien in een .ts file, return-focus. Zou zomaar kunnen dat we dat niet meer nodig hebben met de `<dialog>`.

**Comment by LukeVerlaan** [comment]
Er zijn nog 3 tests die falen, `.should("have.focus")`. Die focus is er wel.
Die errors uit de screenshot hierboven moeten ook nog worden opgelost

Misschien kan @tfrijsewijk hier even naar kijken als hij tijd heeft. Die ziet vaak wat sneller de oorzaak.

En de linterror is volgens mij een comment in scss die op een verkeerde plaats staat

**Comment by hansgrimm** [comment]
recent toegevoegde comment verwijderd, maar wat daar verkeerd aan kan zijn vat ik niet helemaal ..

**Comment by tfrijsewijk** [comment]
Ik zie dat je idd een comment hebt verwijderd, ik denk alleen niet dat dat iets uit gaat maken. Maar dat vermoeden heb jij ook.

Een aantal tips:

## Lokaal linten
Je kan lokaal linten met `yarn lint`, dan hoef je niet op de buildstraat te wachten

## Build debuggen
Je kan het buildlog openen door op de status indicator van de commit te klikken:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/5a278dc8-94cc-46aa-bc18-8bfab21e9b70)

Dit kan naast het oranje bolletje nog een groen vinkje of rood kruisje zijn. Bij het activeren van de status activator krijg je een widget met een samenvatting te zien:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/a82d5bb4-282c-4e57-83ac-242e69dcc705)

Als je vervolgens op Details ga je naar de detail pagina van GitHub:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/45fc1393-5ffd-4556-af00-0a8011fbd4c7)

Door op de link "The build" te klikken ga je naar de build detail pagina in Travis. Hier kan je het buildlog uitvoerig inspecteren:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/140fc5ea-ef73-4e13-884a-8782d37f8053)

De errors staan meestal onderaan, want de build stopt na een error. Uitzonderingen daargelaten natuurlijk ;-)

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/c79ec7a0-daaa-4748-9d23-b2ba12a52992)

Het probleem is dat een line of code te lang is:

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/a2b553d0-9367-4e8e-9929-6030f61640b3)

![image](https://github.com/dso-toolkit/dso-toolkit/assets/9696261/3f155142-3293-47d7-b6d2-9df1890ee450)

## Prettier format

Met `yarn format` kan je prettier alle files laten formatten. In principe is dit nooit slecht, maar altijd even goed opletten bij het stagen van je changes.

## Format on Save

Je kan je IDE instellen om bij het saven van je file automatisch met Prettier te laten formatten.

**Comment by hansgrimm** [comment]
![Scherm­afbeelding 2023-08-14 om 10 08 37](https://github.com/dso-toolkit/dso-toolkit/assets/506527/f21c9914-c87b-456e-8570-50c537790fed)

**Comment by LukeVerlaan** [comment]
Ah ja, dat ben ik geweest

**Comment by tfrijsewijk** [comment]
Ik ga kijken!

**Comment by tfrijsewijk** [comment]
Ik heb de foutmelding gefixed: https://github.com/dso-toolkit/dso-toolkit/commit/adc877c64830bc839416f51c37c8e9ee30ac792b

Ik weet waar het door komt, maar wellicht gebruik ik de verkeerde termen: Het komt erop neer dat een module (file) in de `storybook` package een import naar `@dso-toolkit/core` deed waardoor alle global/ambient typings "actief" werden.

De import die dit deed was `import { DsoModalController } from "@dso-toolkit/core";` Doordat de Controllers zijn verwijderd was er geen enkele `import "@dso-toolkit/core";` meer waardoor de global/ambient typings niet meer werden geladen.

Ik denk dat `tsconfig.json/compilerOptions/typings` de juiste manier is omdeze typings weer te activeren, maar we moeten het even in de gaten houden.

De Modal tests falen op:

- "should focus on initialFocus selector when given".
- "should have keyboard accessible body container".

Ik denk dat deze tests falen op een fundamenteel verschil. Als ik https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog#accessibility_considerations goed begrijp regelt `dialog` voor ons de focus trap. Volgens mij moeten we `focus-trap` niet gebruiken.

Hier nog wat losse opmerkingen waar ik het graag met jullie over wil hebben:
- is `role="document"` nog nodig?
- is `body.dso-modal-open` nodig nodig?

**Comment by hansgrimm** [comment]
De `<dialog>` heeft een impliciete `role="dialog"`, dus die is in principe overbodig. echter, voor oudere screenreaders die nog niet up to date zijn met het element, lan het handig zijn om die role er toch in te laten zitten; kan geen kwaad.
De `role="document"` is bedoeld om stukken leestekst als zodanig te laten voorlezen binnen een widget of application. 
zie https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/document_role (die gebruikt een `<dialog>` als voorbeeld, heel toepasselijk ;)

`body.dso-modal-open` zet de body op `overflow: hidden`, zodat die niet kan scrollen zolang de `<dialog>` geopend is. aangezien de browser dat niet standaard doet, is die nog steeds nuttig.

**Comment by tfrijsewijk** [comment]
@LukeVerlaan @hansgrimm: Hij is gemerged! Dank voor de inzet en het geduld 🖖

**Comment by hansgrimm** [comment]
@tfrijsewijk dat is graag gedaan! Was een mooie en interessante klus! 👌
